File: shared/src/commonMain/kotlin/space/denhan/extract/OpenAiClient.kt
package space.denhan.extract

import io.ktor.client.HttpClient
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import kotlinx.serialization.json.addJsonObject

/**
 * Structured extraction against OpenAI's Responses API.
 *
 * Sends TEXT, never the image. The screenshot is recognised on-device and the
 * user reviews and redacts the recognised lines before anything is transmitted,
 * which is what makes the redaction meaningful: there is no original image on
 * the wire for a missed region to leak through.
 *
 * See product-spec.md sections 8.2bis and 8.4.
 */
class OpenAiClient(
    private val http: HttpClient,
    private val apiKeyProvider: () -> String?,
    private val model: String = DEFAULT_MODEL,
    private val baseUrl: String = "https://api.openai.com/v1",
) {

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    suspend fun extract(
        text: String,
        todayIso: String,
        locale: String,
    ): Result<ExtractedFields> {
        val key = apiKeyProvider() ?: return Result.failure(ExtractionException(ExtractionError.NoApiKey))

        val response: HttpResponse = try {
            http.post("$baseUrl/responses") {
                header("Authorization", "Bearer $key")
                contentType(ContentType.Application.Json)
                setBody(requestBody(text, todayIso, locale).toString())
            }
        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e
        } catch (_: Throwable) {
            return Result.failure(ExtractionException(ExtractionError.NoNetwork))
        }

        val raw = response.bodyAsText()

        if (!response.status.isSuccess()) {
            return Result.failure(ExtractionException(parseError(response, raw)))
        }

        return parseSuccess(raw)
    }

    private fun requestBody(text: String, todayIso: String, locale: String): JsonObject = buildJsonObject {
        put("model", model)

        // Reasoning tokens bill as output tokens and are the largest line item in
        // a request that needs no reasoning. Left at the default this quietly
        // triples the cost of every extraction.
        putJsonObject("reasoning") { put("effort", "low") }

        // A truncated strict-JSON response is unparseable, not partially valid,
        // so the cap has to be paired with a status check below.
        put("max_output_tokens", 800)

        putJsonArray("input") {
            addJsonObject {
                put("role", "system")
                put("content", ExtractionSchema.systemPrompt(todayIso, locale))
            }
            addJsonObject {
                put("role", "user")
                put("content", text)
            }
        }

        // Responses API spelling. `response_format` is the Chat Completions
        // equivalent and is the most common porting mistake; `name` is a sibling
        // of `schema`, not nested under a wrapper.
        putJsonObject("text") {
            putJsonObject("format") {
                put("type", "json_schema")
                put("name", ExtractionSchema.NAME)
                put("strict", true)
                put("schema", Json.parseToJsonElement(ExtractionSchema.json))
            }
        }
    }

    private fun parseError(response: HttpResponse, raw: String): ExtractionError {
        val error = runCatching {
            json.parseToJsonElement(raw).jsonObject["error"]?.jsonObject
        }.getOrNull()

        return ErrorMapper.map(
            status = response.status.value,
            errorCode = error?.get("code")?.jsonPrimitive?.contentOrNullSafe(),
            errorType = error?.get("type")?.jsonPrimitive?.contentOrNullSafe(),
            retryAfterSeconds = response.headers["Retry-After"]?.toIntOrNull(),
        )
    }

    private fun parseSuccess(raw: String): Result<ExtractedFields> {
        val root = runCatching { json.parseToJsonElement(raw).jsonObject }.getOrNull()
            ?: return Result.failure(ExtractionException(ExtractionError.Malformed("not json")))

        // Check status before parsing: an incomplete response holds truncated
        // JSON that will not parse and must not be treated as a parse bug.
        if (root["status"]?.jsonPrimitive?.contentOrNullSafe() == "incomplete") {
            return Result.failure(ExtractionException(ExtractionError.Truncated))
        }

        val content = root["output"]?.jsonArray
            ?.firstOrNull()?.jsonObject
            ?.get("content")?.jsonArray
            ?.firstOrNull()?.jsonObject

        // A refusal does not honour the schema at all.
        content?.get("refusal")?.jsonPrimitive?.contentOrNullSafe()?.let {
            return Result.failure(ExtractionException(ExtractionError.Refused(it)))
        }

        val payload = content?.get("text")?.jsonPrimitive?.contentOrNullSafe()
            ?: root["output_text"]?.jsonPrimitive?.contentOrNullSafe()
            ?: return Result.failure(ExtractionException(ExtractionError.Malformed("no output text")))

        return runCatching { json.decodeFromString<ExtractedFields>(payload) }
            .fold(
                onSuccess = { Result.success(it) },
                onFailure = { Result.failure(ExtractionException(ExtractionError.Malformed(it.message))) },
            )
    }

    companion object {
        /**
         * Current-generation budget tier. Not gpt-5-nano: cheaper on paper but
         * deprecated with a shutdown date and no Responses API support.
         */
        const val DEFAULT_MODEL = "gpt-5.6-luna"
    }
}

class ExtractionException(val error: ExtractionError) : Exception(error.messageVi)

private fun kotlinx.serialization.json.JsonPrimitive.contentOrNullSafe(): String? =
    if (this is kotlinx.serialization.json.JsonNull) null else content

private fun kotlinx.serialization.json.JsonObjectBuilder.putJsonArray(
    key: String,
    builder: kotlinx.serialization.json.JsonArrayBuilder.() -> Unit,
) {
    put(key, buildJsonArray(builder))
}

File: lib/extract/openai_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import 'extraction_error.dart';
import 'extraction_schema.dart';

/// Structured extraction against OpenAI's Responses API.
///
/// Sends TEXT, never the image. The screenshot is recognised on-device and the
/// user reviews and redacts the recognised lines before anything is
/// transmitted, which is what makes the redaction meaningful: there is no
/// original image on the wire for a missed region to leak through.
///
/// See product-spec.md sections 8.2bis and 8.4.
class OpenAiClient {
  /// Current-generation budget tier. Not gpt-5-nano: cheaper on paper but
  /// deprecated with a shutdown date and no Responses API support.
  static const String defaultModel = 'gpt-5.6-luna';

  final http.Client _http;
  final String? Function() _apiKeyProvider;

  final String model;
  final String baseUrl;

  OpenAiClient({
    required http.Client httpClient,
    required String? Function() apiKeyProvider,
    this.model = defaultModel,
    this.baseUrl = 'https://api.openai.com/v1',
  }) : _http = httpClient,
       // ignore: prefer_initializing_formals
       _apiKeyProvider = apiKeyProvider;

  /// Returns the parsed fields, or throws [ExtractionException].
  ///
  /// Every failure path carries an [ExtractionError] that already knows whether
  /// retrying could help, so no caller has to re-derive that from a status code.
  Future<ExtractedFields> extract({
    required String text,
    required String todayIso,
    required String locale,
  }) async {
    final key = _apiKeyProvider();
    if (key == null || key.isEmpty) {
      throw const ExtractionException(NoApiKey());
    }

    late final http.Response response;
    try {
      response = await _http.post(
        Uri.parse('$baseUrl/responses'),
        headers: {
          'Authorization': 'Bearer $key',
          'Content-Type': 'application/json',
        },
        body: jsonEncode(_requestBody(text, todayIso, locale)),
      );
    } on http.ClientException {
      throw const ExtractionException(NoNetwork());
    } on SocketException {
      throw const ExtractionException(NoNetwork());
    } on HandshakeException {
      throw const ExtractionException(NoNetwork());
    }

    final body = _decodeBody(response);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ExtractionException(_parseError(response, body));
    }

    return _parseSuccess(body);
  }

  /// Decodes the body as UTF-8 explicitly.
  ///
  /// `http.Response.body` picks its codec from the response's `charset`
  /// parameter and falls back to **latin-1** when the server omits one. JSON is
  /// UTF-8 by RFC 8259, so that fallback mangles every Vietnamese character in
  /// a reply from a server that was merely terse about its headers, and the
  /// damage looks like a bad extraction rather than a decoding bug.
  String _decodeBody(http.Response response) {
    try {
      return utf8.decode(response.bodyBytes);
    } on FormatException {
      // Not valid UTF-8 at all: report it as malformed rather than pretending
      // some other encoding was meant.
      throw const ExtractionException(Malformed('response was not UTF-8'));
    }
  }

  Map<String, dynamic> _requestBody(
    String text,
    String todayIso,
    String locale,
  ) {
    return {
      'model': model,

      // Reasoning tokens bill as output tokens and are the largest line item
      // in a request that needs no reasoning. Left at the default this quietly
      // triples the cost of every extraction.
      'reasoning': {'effort': 'low'},

      // A truncated strict-JSON response is unparseable, not partially valid,
      // so the cap has to be paired with the status check in _parseSuccess.
      'max_output_tokens': 800,

      'input': [
        {
          'role': 'system',
          'content': ExtractionSchema.systemPrompt(todayIso, locale),
        },
        {'role': 'user', 'content': text},
      ],

      // Responses API spelling. `response_format` is the Chat Completions
      // equivalent and is the most common porting mistake; `name` is a sibling
      // of `schema`, not nested under a wrapper.
      'text': {
        'format': {
          'type': 'json_schema',
          'name': ExtractionSchema.name,
          'strict': true,
          'schema': jsonDecode(ExtractionSchema.json),
        },
      },
    };
  }

  ExtractionError _parseError(http.Response response, String body) {
    Map<String, dynamic>? error;
    try {
      final root = jsonDecode(body);
      if (root is Map<String, dynamic>) {
        final candidate = root['error'];
        if (candidate is Map<String, dynamic>) error = candidate;
      }
    } on FormatException {
      // A non-JSON error body is normal for a gateway failure. The status
      // still classifies it.
    }

    return ErrorMapper.map(
      response.statusCode,
      errorCode: error?['code'] as String?,
      errorType: error?['type'] as String?,
      retryAfterSeconds: int.tryParse(response.headers['retry-after'] ?? ''),
    );
  }

  ExtractedFields _parseSuccess(String raw) {
    final Object? decoded;
    try {
      decoded = jsonDecode(raw);
    } on FormatException {
      throw const ExtractionException(Malformed('not json'));
    }

    if (decoded is! Map<String, dynamic>) {
      throw const ExtractionException(Malformed('not json'));
    }

    // Check status before parsing: an incomplete response holds truncated JSON
    // that will not parse and must not be reported as a parse bug.
    if (decoded['status'] == 'incomplete') {
      throw const ExtractionException(Truncated());
    }

    final content = _firstContent(decoded);

    // A refusal does not honour the schema at all.
    final refusal = content?['refusal'];
    if (refusal is String) {
      throw ExtractionException(Refused(refusal));
    }

    final payload =
        (content?['text'] as String?) ?? (decoded['output_text'] as String?);
    if (payload == null) {
      throw const ExtractionException(Malformed('no output text'));
    }

    try {
      final fields = jsonDecode(payload);
      if (fields is! Map<String, dynamic>) {
        throw const FormatException('payload is not an object');
      }
      return ExtractedFields.fromJson(fields);
    } on ExtractionException {
      rethrow;
    } on FormatException catch (e) {
      throw ExtractionException(Malformed(e.message));
    } on TypeError catch (e) {
      // The reply parsed as JSON but had a field of the wrong shape.
      throw ExtractionException(Malformed(e.toString()));
    }
  }

  Map<String, dynamic>? _firstContent(Map<String, dynamic> root) {
    final output = root['output'];
    if (output is! List || output.isEmpty) return null;

    final first = output.first;
    if (first is! Map<String, dynamic>) return null;

    final content = first['content'];
    if (content is! List || content.isEmpty) return null;

    final firstContent = content.first;
    return firstContent is Map<String, dynamic> ? firstContent : null;
  }
}

class ExtractionException implements Exception {
  final ExtractionError error;

  const ExtractionException(this.error);

  String get message => error.message;

  bool get retryable => error.retryable;

  @override
  String toString() =>
      'ExtractionException(${error.runtimeType}): ${error.message}';
}
